### PR TITLE
refactor: Flatten update lookup tasks

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Collect all dep fetches for all package files for all managers into the single array
- Process it concurrently (with the limit of simulataneous runs)
- Adjust error handling to behave similarly to current functionality

## Context

- Ref: #6575

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
This one should be tested well

**Comment by zharinov** [comment]
I haven't formalized my thoughts yet, but here we may want the single queue to be cancelable. Because, even we get `ExternalHostError`, fetching continues to some degree for both current and this new implementations.

**Comment by zharinov** [comment]
@rarkins Could you provide a hint of how lookup error handling should work? Given we have flat list of lookup tasks performed concurrently, it's not clear how to deal with errors:
- Should some errors stop processing completely? (i.e., for all hosts and for all managers)
- Should some errors stop further requests to particular host, while others are processed as usual?
- Should we log/collect other types of errors?

File: lib/util/promises.ts
Comment (viceice) [inline lib/util/promises.ts:31]: We need to be careful, `AggregateError`  is a javascript build-in too since node v15[^1].

[^1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError

File: lib/workers/repository/process/fetch.ts
Comment (viceice) [inline lib/workers/repository/process/fetch.ts:24]: prefer interfaces if possible

File: lib/workers/repository/process/fetch.spec.ts
Comment (viceice) [inline lib/workers/repository/process/fetch.spec.ts:1]: There's a failing snapshot 🤔

File: lib/workers/repository/process/fetch.ts
Comment (viceice) [inline lib/workers/repository/process/fetch.ts:107]: are we sure `warnings` was never initialized before? otherwise the flat merge below will overwrite existing values with this.

File: lib/workers/repository/process/fetch.ts
Comment (viceice) [inline lib/workers/repository/process/fetch.ts:168]: so this is the 5by5 concurrency we've done before?

**Review by viceice — COMMENTED** [review]
I think we should cancel early on specific errors 🤔 it looks like we now do all lookups and doing a lot more failing requests against datasources than before.

**Comment by rarkins** [comment]
@viceice I'm not sure we can easily cancel when errors are encountered. Requests to the same host may already be in-progress, etc. Logic is definitely easier if we try all/fail many. Are you after the "speed" benefit mostly?

**Comment by rarkins** [comment]
So we'll have a flat list of X lookups, and at some point one of them returns a critical error. Your suggestion is to abort all the other lookups which haven't yet started? I don't think we can reliably abort lookups which are already in motion.

**Comment by viceice** [comment]
not cancel in flight but waiting ones. p.map will call in groups by max parallelism. so we should stop processing when a group fails.

**Comment by rarkins** [comment]
My plans were to later have full concurrency at the application layer here and do per host limiting at the http layer

File: lib/util/promises.ts
Comment (zharinov) [reply lib/util/promises.ts:31]: In this context, we care only about the `AggregateError` variant thrown by `p-map`/`p-all`

File: lib/workers/repository/process/fetch.ts
Comment (zharinov) [reply lib/workers/repository/process/fetch.ts:168]: Not exactly, but somewhat similar.

Before: `N managers ✖️ 5 package files ✖️ 5 deps`
After: `5 ✖️ 5 deps`

File: lib/workers/repository/process/fetch.ts
Comment (zharinov) [reply lib/workers/repository/process/fetch.ts:109]: Since it's the `catch` path, it's unlikely we've pushed any warnings before

File: lib/workers/repository/process/fetch.ts
Comment (rarkins) [reply lib/workers/repository/process/fetch.ts:169]: I would like to have default concurrent request limits for every host if we don't already have it. What would be a reasonable value?

Comment (zharinov) [reply lib/workers/repository/process/fetch.ts:169]: I think it's hard to enforce this only for lookups. We could set default global concurrency limits [here](https://github.com/renovatebot/renovate/blob/38d85ce23fb830ff17fac265f4d72b244c648d05/lib/util/http/host-rules.ts#L220), as the separate PR (current is no limits).

**Comment by zharinov** [comment]
@viceice While it's reasonable to prevent further requests to the same failed host, we don't have it for now anyway, so let's maybe consider it as the new feature with separate PR.

**Comment by rarkins** [comment]
@zharinov what changes does this currently make to concurrency? Do we need to consider adding per-host limits first?

**Comment by zharinov** [comment]
It makes 25 concurrent requests to same host possible, instead of 5. Seems like we may hit 429, so I think yes, let's change our defaults first.

**Comment by zharinov** [comment]
- Blocked by #27621

File: lib/workers/repository/process/fetch.ts
Comment (viceice) [inline lib/workers/repository/process/fetch.ts:169]: use promise.all or allsettled, we don't need pMap here if we do infinite parallel calls.

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs rate limiting first

**Review by viceice — DISMISSED** [review]
🤷‍♂️  needs extensive testing

**Comment by rarkins** [comment]
@zharinov needs deconflicting

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]
🤷‍♂️

File: lib/workers/repository/process/fetch.ts
Comment (viceice) [inline lib/workers/repository/process/fetch.ts:159]: we should use all settled, so we don't have lingering tasks because all is breaking early on errors and don't cancel remaining tasks

Comment (zharinov) [reply lib/workers/repository/process/fetch.ts:159]: Cool alternative, but it won't add any benefit, because we use custom error handling here anyway (never throws)

**Review by viceice — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
This causes OOM with repo cache and regex managers, so it's getting reverted.
- https://github.com/renovatebot/renovate/discussions/30906

**Pull request merged**
